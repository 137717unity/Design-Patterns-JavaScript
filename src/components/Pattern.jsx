import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Light as SyntaxHighlighter } from 'react-syntax-highlighter';
import js from 'react-syntax-highlighter/dist/languages/hljs/javascript';
import styleLight from '../styles/hljs/hljs.light';
import styleDark from '../styles/hljs/hljs.dark';
import { patterns } from '../static/patterns';
import { getMode } from '../selectors';

SyntaxHighlighter.registerLanguage('javascript', js);

const StyledPattern = styled.div`
  color: ${props => props.theme.text};

  h2,
  h3 {
    color: ${props => props.theme.header};
    margin-top: 2.5rem;
  }
`;

const Memo = styled.span`
  color: ${props => props.theme.ATLANTIS};
  display: block;
`;

const StyledLink = styled(Link)`
  border-bottom: 1px solid ${props => props.theme.CRIMSON};
  color: ${props => props.theme.CRIMSON};
  display: inline-block;
  margin-top: 2rem;
  text-decoration: none;

  &:hover {
    border-bottom: 1px solid transparent;
  }
`;

const Pattern = ({ match, mode }) => {
  const {
    params: { id }
  } = match;

  const pattern = patterns.filter(item => item.id === id)[0];

  const style = mode === 'dark' ? styleDark : styleLight;

  return (
    <StyledPattern>
      <StyledLink to="/patterns">&larr; Back to Patterns List</StyledLink>
      {pattern && (
        <React.Fragment>
          <h2>{pattern.name}</h2>
          <p>
            <Memo>Type:</Memo>
            {pattern.type} pattern
          </p>
          <p>
            <Memo>Description:</Memo>
            {`This pattern ${pattern.hint}.`}
          </p>

          <h3>ES5</h3>
          <SyntaxHighlighter language="javascript" style={style}>
            {pattern.codeES5}
          </SyntaxHighlighter>

          <h3>ES6</h3>
          <SyntaxHighlighter language="javascript" style={style}>
            {pattern.codeES6}
          </SyntaxHighlighter>
        </React.Fragment>
      )}
    </StyledPattern>
  );
};

Pattern.propTypes = {
  match: PropTypes.object.isRequired,
  mode: PropTypes.string.isRequired
};

export default connect(state => ({
  mode: getMode(state)
}))(Pattern);
