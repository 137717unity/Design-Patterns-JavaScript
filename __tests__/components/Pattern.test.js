import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import 'jest-styled-components';
import Pattern from '../../src/components/Pattern';

const mockStore = configureMockStore();

describe('REFERENCE - Pattern component', () => {
  it('renders the individual pattern (Singleton) details in DARK mode', () => {
    const store = mockStore({
      mode: 'dark'
    });
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={['/patterns/singleton']}>
          <Provider store={store}>
            <Pattern match={{ params: { id: 'singleton' } }} />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  xit('renders the individual pattern (Singleton) details in LIGHT mode', () => {
    const store = mockStore({
      mode: 'light'
    });
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={['/patterns/singleton']}>
          <Provider store={store}>
            <Pattern match={{ params: { id: 'singleton' } }} />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
