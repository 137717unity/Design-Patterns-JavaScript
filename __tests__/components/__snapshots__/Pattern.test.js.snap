// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REFERENCE - Pattern component renders the individual pattern (Singleton) details in DARK mode 1`] = `
.c0 h2,
.c0 h3 {
  margin-top: 2.5rem;
}

.c2 {
  display: block;
  margin-bottom: 8px;
  padding: 4px;
  text-transform: uppercase;
}

.c3 {
  text-transform: capitalize;
}

.c1 {
  border-bottom: 1px solid;
  display: inline-block;
  margin-top: 2rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:hover {
  border-bottom: 1px solid transparent;
}

<div
  className="c0"
>
  <a
    className="c1"
    href="/patterns"
    onClick={[Function]}
  >
    ← Back to Patterns List
  </a>
  <h2>
    Singleton
  </h2>
  <p>
    <span
      className="c2"
    >
      Type:
    </span>
    <span
      className="c3"
    >
      creational
       pattern
    </span>
  </p>
  <p>
    <span
      className="c2"
    >
      Definition:
    </span>
    Ensure a class has only one instance and provide a global point of access to it.
  </p>
  <p>
    <span
      className="c2"
    >
      Use when…
    </span>
    …
    there must by only one instance of a class
    .
  </p>
  <h3>
    ES5
  </h3>
  <pre
    style={
      Object {
        "background": "#282828",
        "border": "1px solid #555",
        "color": "#ebdbb2",
        "display": "block",
        "overflowX": "auto",
        "padding": "1em",
      }
    }
  >
    <code>
      <span
        style={
          Object {
            "color": "#8ec07c",
          }
        }
      >
        <span
          style={
            Object {
              "color": "#fb4934",
            }
          }
        >
          function
        </span>
         
        <span
          style={
            Object {
              "color": "#83a598",
            }
          }
        >
          Person
        </span>
        (
        <span
          style={
            Object {
              "color": "#fabd2f",
            }
          }
        />
        ) 
      </span>
      {
  
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        if
      </span>
       (
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        typeof
      </span>
       Person.instance === 
      <span
        style={
          Object {
            "color": "#b8bb26",
          }
        }
      >
        'object'
      </span>
      ) 
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        return
      </span>
       Person.instance;

  Person.instance = 
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        this
      </span>
      ;

  
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        return
      </span>
       
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        this
      </span>
      ;
}


      <span
        style={
          Object {
            "color": "#83a598",
          }
        }
      >
        module
      </span>
      .exports = Person;
    </code>
  </pre>
  <h3>
    ES6
  </h3>
  <pre
    style={
      Object {
        "background": "#282828",
        "border": "1px solid #555",
        "color": "#ebdbb2",
        "display": "block",
        "overflowX": "auto",
        "padding": "1em",
      }
    }
  >
    <code>
      <span
        style={
          Object {
            "color": "#8ec07c",
          }
        }
      >
        <span
          style={
            Object {
              "color": "#fb4934",
            }
          }
        >
          class
        </span>
         
        <span
          style={
            Object {
              "color": "#83a598",
            }
          }
        >
          Person
        </span>
         
      </span>
      {
  
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        constructor
      </span>
      () {
    
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        if
      </span>
       (
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        typeof
      </span>
       Person.instance === 
      <span
        style={
          Object {
            "color": "#b8bb26",
          }
        }
      >
        'object'
      </span>
      ) {
      
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        return
      </span>
       Person.instance;
    }

    Person.instance = 
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        this
      </span>
      ;

    
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        return
      </span>
       
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        this
      </span>
      ;
  }
}


      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        export
      </span>
       
      <span
        style={
          Object {
            "color": "#fb4934",
          }
        }
      >
        default
      </span>
       Person;
    </code>
  </pre>
</div>
`;
